### PR TITLE
[ios, build] Updates for Xcode 9.3

### PR DESCRIPTION
**Pull request by friedbunny**
Enables new warnings in Xcode 9.3 and fixes some resulting issues.

In particular, `CLANG_WARN_OBJC_IMPLICIT_RETAIN_SELF` prompted me to weakify some variables that were being captured by blocks. @fabian-guerra I know `MGLMapSnapshotter` memory management has been an issue, so I’d appreciate your eyes on that. 🙇 

The memory graph looks flat and the integration tests pass, in any case.

screen shot 2018-04-24 at 8 22 07 pm
_Loading the snapshot view controller repeatedly in iosapp on an iPhone X._

/cc @1ec5 @fabian-guerra @julianrex

### CODE REVIEW

File: platform/ios/src/MGLMapView.mm [hunk 1]
```diff
@@ -4583,10 +4585,11 @@ - (void)calloutViewWillAppear:(UIView <MGLCalloutView> *)calloutView
                                  UIViewAnimationOptionBeginFromCurrentState)
                      animations:^
      {
+         __typeof__(self) strongSelf = weakSelf;
```
Comment (friedbunny) [inline platform/ios/src/MGLMapView.mm]: @1ec5 you appear to be the original author here — does this fit your expectation?

Comment (1ec5) [reply platform/ios/src/MGLMapView.mm]: Yes. However, if `weakSelf` is already `nil`, then dereferencing it below will crash. Either return early or turn `initialImplicitCalloutViewOffset` into a property whose getter message you can send more safely to `nil`.

File: platform/ios/ios.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -3470,13 +3466,15 @@
...
 				CLANG_WARN_CONSTANT_CONVERSION = YES;
+				CLANG_WARN_DEPRECATED_OBJC_IMPLEMENTATIONS = YES;
 				CLANG_WARN_DIRECT_OBJC_ISA_USAGE = YES_ERROR;
 				CLANG_WARN_DOCUMENTATION_COMMENTS = YES;
 				CLANG_WARN_EMPTY_BODY = YES;
 				CLANG_WARN_ENUM_CONVERSION = YES;
 				CLANG_WARN_INFINITE_RECURSION = YES;
 				CLANG_WARN_INT_CONVERSION = YES;
 				CLANG_WARN_NON_LITERAL_NULL_CONVERSION = YES;
+				CLANG_WARN_OBJC_IMPLICIT_RETAIN_SELF = YES;
```
Comment (1ec5) [inline platform/ios/ios.xcodeproj/project.pbxproj]: Unless the same change is also made to macos.xcodeproj, I and folks on the core team may accidentally introduce issues that would be caught by this warning.

Comment (friedbunny) [reply platform/ios/ios.xcodeproj/project.pbxproj]: Ah, an excellent reminder to also `make xproj` this PR — will do.

File: platform/ios/src/MGLMapView.mm (hunk 1 above)
Comment (friedbunny) [reply platform/ios/src/MGLMapView.mm]: Went with “return early” in https://github.com/mapbox/mapbox-gl-native/pull/11771/commits/f194ddd3d21ac9996f94a3d4a7bdc995a0404504.

File: platform/ios/ios.xcodeproj/project.pbxproj (hunk 2 above)
Comment (friedbunny) [reply platform/ios/ios.xcodeproj/project.pbxproj]: Rolled this into 2b378c7.

File: platform/darwin/src/MGLMapSnapshotter.mm [hunk 3]
```diff
@@ -145,9 +144,10 @@ - (void)startWithQueue:(dispatch_queue_t)queue completionHandler:(MGLMapSnapshot
         }
         _snapshotCallback = NULL;
     });
-    dispatch_async(queue, ^{
-        _mbglMapSnapshotter->snapshot(_snapshotCallback->self());
 
+    dispatch_async(queue, ^{
+        __typeof__(self) strongSelf = weakSelf;
+        strongSelf->_mbglMapSnapshotter->snapshot(strongSelf->_snapshotCallback->self());
```
Comment (fabian-guerra) [inline platform/darwin/src/MGLMapSnapshotter.mm]: Thanks for fixing this.

File: platform/ios/ios.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -53,7 +53,6 @@
 		3510FFEC1D6D9C7A00F413B2 /* NSComparisonPredicate+MGLAdditions.mm in Sources */ = {isa = PBXBuildFile; fileRef = 3510FFE91D6D9C7A00F413B2 /* NSComparisonPredicate+MGLAdditions.mm */; };
 		3510FFED1D6D9C7A00F413B2 /* NSComparisonPredicate+MGLAdditions.mm in Sources */ = {isa = PBXBuildFile; fileRef = 3510FFE91D6D9C7A00F413B2 /* NSComparisonPredicate+MGLAdditions.mm */; };
 		3510FFF01D6D9D8C00F413B2 /* NSExpression+MGLAdditions.h in Headers */ = {isa = PBXBuildFile; fileRef = 3510FFEE1D6D9D8C00F413B2 /* NSExpression+MGLAdditions.h */; settings = {ATTRIBUTES = (Public, ); }; };
-		3510FFF11D6D9D8C00F413B2 /* NSExpression+MGLAdditions.h in Headers */ = {isa = PBXBuildFile; fileRef = 3510FFEE1D6D9D8C00F413B2 /* NSExpression+MGLAdditions.h */; settings = {ATTRIBUTES = (Public, ); }; };
```
Comment (fabian-guerra) [inline platform/ios/ios.xcodeproj/project.pbxproj]: I don't know why this file keeps popping out. Just make sure it's public.

Comment (friedbunny) [reply platform/ios/ios.xcodeproj/project.pbxproj]: Even now, it seems to exist twice:

screen shot 2018-04-25 at 2 34 56 pm

... will take another look.

Comment (friedbunny) [reply platform/ios/ios.xcodeproj/project.pbxproj]: This should be taken care of in: https://github.com/mapbox/mapbox-gl-native/pull/11771/commits/650adae9fc3f3c73c6de446ebf049c99ca1abbf3

File: platform/darwin/src/MGLMapSnapshotter.mm [hunk 5]
```diff
@@ -145,9 +144,10 @@ - (void)startWithQueue:(dispatch_queue_t)queue completionHandler:(MGLMapSnapshot
         }
         _snapshotCallback = NULL;
     });
-    dispatch_async(queue, ^{
-        _mbglMapSnapshotter->snapshot(_snapshotCallback->self());
 
+    dispatch_async(queue, ^{
+        __typeof__(self) strongSelf = weakSelf;
```
Comment (1ec5) [inline platform/darwin/src/MGLMapSnapshotter.mm]: Also return early here before dereferencing `strongSelf`.

File: platform/macos/app/MapDocument.m [hunk 6]
```diff
@@ -185,8 +184,11 @@ - (IBAction)takeSnapshot:(id)sender {
     options.zoomLevel = self.mapView.zoomLevel;
     
     // Create and start the snapshotter
-    snapshotter = [[MGLMapSnapshotter alloc] initWithOptions:options];
-    [snapshotter startWithCompletionHandler:^(MGLMapSnapshot *snapshot, NSError *error) {
+    __weak __typeof__(self) weakSelf = self;
+    _snapshotter = [[MGLMapSnapshotter alloc] initWithOptions:options];
+    [_snapshotter startWithCompletionHandler:^(MGLMapSnapshot *snapshot, NSError *error) {
+        __typeof__(self) strongSelf = weakSelf;
```
Comment (1ec5) [inline platform/macos/app/MapDocument.m]: Return early before dereferencing `strongSelf` below. It’s definitely possible to close the window before the snapshot is complete, thereby causing a crash.

File: platform/darwin/src/MGLMapSnapshotter.mm (hunk 5 above)
Comment (friedbunny) [reply platform/darwin/src/MGLMapSnapshotter.mm]: I think I may be misunderstanding: this doesn’t explicitly deference anything right now — should it?

Comment (1ec5) [reply platform/darwin/src/MGLMapSnapshotter.mm]: The line below dereferences `strongSelf`. It’s important to check for a `nil` `strongSelf`, not a `nil` `weakSelf`.

File: platform/darwin/src/MGLMapSnapshotter.mm [hunk 7]
```diff
@@ -146,6 +146,9 @@ - (void)startWithQueue:(dispatch_queue_t)queue completionHandler:(MGLMapSnapshot
     });
 
     dispatch_async(queue, ^{
+        if (!weakSelf) {
+            return;
```
Comment (1ec5) [inline platform/darwin/src/MGLMapSnapshotter.mm]: This introduces a race condition in which `weakSelf` could potentially be nilled out between here and the next line, where `weakSelf` (now `nil`) is assigned to `strongSelf`.

File: platform/darwin/src/MGLMapSnapshotter.mm (hunk 5 above)
Comment (friedbunny) [reply platform/darwin/src/MGLMapSnapshotter.mm]: Ah, alright — this is done in https://github.com/mapbox/mapbox-gl-native/pull/11771/commits/fc8be4becd1b0c4f609b35f058200e0c2b1d4d64. Chatted with @julianrex a bit about this (and, separately, the potentially problematic block/ivar interactions in this snapshot code) and he agreed with your suggestion. Thanks to you both for walking me through this. 🙇

File: platform/ios/src/MGLMapView.mm [hunk 8]
```diff
@@ -4583,10 +4585,16 @@ - (void)calloutViewWillAppear:(UIView <MGLCalloutView> *)calloutView
                                  UIViewAnimationOptionBeginFromCurrentState)
                      animations:^
      {
+         if ( ! weakSelf)
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: Check `strongSelf` instead of `weakSelf` here as well.

File: platform/macos/app/MapDocument.m [hunk 9]
```diff
@@ -185,17 +184,23 @@ - (IBAction)takeSnapshot:(id)sender {
     options.zoomLevel = self.mapView.zoomLevel;
     
     // Create and start the snapshotter
-    snapshotter = [[MGLMapSnapshotter alloc] initWithOptions:options];
-    [snapshotter startWithCompletionHandler:^(MGLMapSnapshot *snapshot, NSError *error) {
+    __weak __typeof__(self) weakSelf = self;
+    _snapshotter = [[MGLMapSnapshotter alloc] initWithOptions:options];
+    [_snapshotter startWithCompletionHandler:^(MGLMapSnapshot *snapshot, NSError *error) {
+        if (!weakSelf) {
```
Comment (1ec5) [inline platform/macos/app/MapDocument.m]: And here.

**Comment by mb12** [comment]
@friedbunny and @1ec5 Does upgrade to newer version of xcode require any recompilations of dependencies under mason_packages? If that is the case, would it be possible to increase the version number of those recompiled libs.    https://github.com/mapbox/mapbox-gl-native/issues/11557

**Comment by friedbunny** [comment]
> Does upgrade to newer version of xcode require any recompilations of dependencies under mason_packages?

@mb12 So far as I’m aware, no — the changes in this PR are to the iOS and macOS projects in this repo, while [mason](https://github.com/mapbox/mason) compilation is handled elsewhere.

**Pull request merged**